File: sdks/typescript/src/observability/otlp.ts
/**
 * OTLP (OpenTelemetry Protocol) bridge for `dcpTelemetry`.
 *
 * Kept in its own file so the OTel SDK imports are lazy: nothing from
 * `@opentelemetry/*` is loaded unless `dcpTelemetry.init({ exporterType: 'otlp' })`
 * is called. That keeps the default install footprint and runtime of
 * `@dcp-ai/sdk` unchanged for the 95% of consumers who never turn OTLP on.
 */

import type { DcpTelemetryConfig, TelemetryEvent } from './telemetry.js';

export interface OtlpHandles {
  /** Record the given telemetry event by projecting it into OTel primitives. */
  handleEvent(event: TelemetryEvent): void;
  /** Flush + close the OTel SDK. Best-effort. */
  shutdown(): Promise<void>;
}

// The `@opentelemetry/*` packages are declared as optional peer dependencies,
// so their types are not guaranteed to be installed at compile time. We keep
// the imports dynamic and the handles untyped (`any`) to avoid dragging their
// type surface into `@dcp-ai/sdk`'s public typings. The runtime behaviour is
// still correct because the OTel SDK enforces its own shape at call time.
/* eslint-disable @typescript-eslint/no-explicit-any */

type HistogramMap = Map<string, { record(value: number, attrs?: Record<string, string | number | boolean>): void }>;
type CounterMap = Map<string, { add(value: number, attrs?: Record<string, string | number | boolean>): void }>;

// Suppress type-only import resolution when the optional OTel packages are
// not installed in the consuming environment. We want this file to compile
// against any TS config, dependency present or not.
// @ts-ignore
type OtelImport = any;

async function tryLoadOtel(): Promise<{
  api: OtelImport;
  sdkTraceNode: OtelImport;
  sdkMetrics: OtelImport;
  traceExporter: OtelImport;
  metricsExporter: OtelImport;
  resources: OtelImport;
  semconv: OtelImport;
}> {
  // Variable specifiers keep TypeScript from statically resolving these imports,
  // which would fail the build when the optional OTel packages are absent.
  const imp = (name: string) => import(/* @vite-ignore */ /* webpackIgnore: true */ name);
  const [api, sdkTraceNode, sdkMetrics, traceExporter, metricsExporter, resources, semconv] = await Promise.all([
    imp('@opentelemetry/api'),
    imp('@opentelemetry/sdk-trace-node'),
    imp('@opentelemetry/sdk-metrics'),
    imp('@opentelemetry/exporter-trace-otlp-http'),
    imp('@opentelemetry/exporter-metrics-otlp-http'),
    imp('@opentelemetry/resources'),
    imp('@opentelemetry/semantic-conventions'),
  ]);
  return { api, sdkTraceNode, sdkMetrics, traceExporter, metricsExporter, resources, semconv };
}

/**
 * Initialise the OTel SDK for traces + metrics and return a small adapter
 * that maps DCP telemetry events to OTel spans and histograms/counters.
 *
 * @throws if any of the optional `@opentelemetry/*` packages are missing.
 *         The error message names the packages to install.
 */
export async function initOtlp(config: DcpTelemetryConfig, sdkVersion: string): Promise<OtlpHandles> {
  let api: OtelImport;
  let sdkTraceNode: OtelImport;
  let sdkMetrics: OtelImport;
  let traceExporter: OtelImport;
  let metricsExporter: OtelImport;
  let resources: OtelImport;
  let semconv: OtelImport;

  try {
    ({ api, sdkTraceNode, sdkMetrics, traceExporter, metricsExporter, resources, semconv } = await tryLoadOtel());
  } catch (err) {
    throw new Error(
      '[DCP-AI Telemetry] OTLP exporter requires the optional OpenTelemetry packages. Install them with:\n' +
        '  npm install @opentelemetry/api @opentelemetry/sdk-trace-node @opentelemetry/sdk-metrics ' +
        '@opentelemetry/exporter-trace-otlp-http @opentelemetry/exporter-metrics-otlp-http ' +
        '@opentelemetry/resources @opentelemetry/semantic-conventions\n' +
        `Underlying error: ${(err as Error).message}`,
    );
  }

  const endpoint = config.otlpEndpoint ?? process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? 'http://localhost:4318';
  const traceUrl = endpoint.replace(/\/+$/, '') + '/v1/traces';
  const metricsUrl = endpoint.replace(/\/+$/, '') + '/v1/metrics';

  const resource = new resources.Resource({
    [semconv.SEMRESATTRS_SERVICE_NAME]: config.serviceName,
    [semconv.SEMRESATTRS_SERVICE_VERSION]: sdkVersion,
  });

  const tracerProvider = new sdkTraceNode.NodeTracerProvider({ resource });
  tracerProvider.addSpanProcessor(
    new sdkTraceNode.BatchSpanProcessor(new traceExporter.OTLPTraceExporter({ url: traceUrl })),
  );
  tracerProvider.register();

  const metricReader = new sdkMetrics.PeriodicExportingMetricReader({
    exporter: new metricsExporter.OTLPMetricExporter({ url: metricsUrl }),
    exportIntervalMillis: config.metricsInterval ?? 15_000,
  });
  const meterProvider = new sdkMetrics.MeterProvider({ resource, readers: [metricReader] });
  api.metrics.setGlobalMeterProvider(meterProvider);

  const tracer = api.trace.getTracer('@dcp-ai/sdk', sdkVersion);
  const meter = api.metrics.getMeter('@dcp-ai/sdk', sdkVersion);

  // Pre-create the histograms + counters we know we emit. Creating them lazily
  // inside handleEvent() would also work, but this is cheaper and matches
  // OTel best practice.
  const histograms: HistogramMap = new Map();
  const counters: CounterMap = new Map();
  const histogram = (name: string, unit = 'ms') => {
    const existing = histograms.get(name);
    if (existing) return existing;
    const h = meter.createHistogram(name, { unit, description: `DCP-AI ${name}` });
    histograms.set(name, h);
    return h;
  };
  const counter = (name: string) => {
    const existing = counters.get(name);
    if (existing) return existing;
    const c = meter.createCounter(name, { description: `DCP-AI ${name}` });
    counters.set(name, c);
    return c;
  };

  // Active OTel spans, keyed by our internal DcpSpan.name + startTime pair.
  // Because dcpTelemetry.emit fires on endSpan (we receive the fully-terminated
  // span), we do not need to track anything here — we create+end the OTel span
  // in one step using startActiveSpan with the measured duration attached as an
  // attribute. That's simpler and faster than maintaining a parallel span map.

  function handleEvent(event: TelemetryEvent): void {
    try {
      if (event.type === 'span') {
        const span = (event as any).span as { name: string; attributes: Record<string, unknown>; status: 'ok' | 'error'; error?: string };
        const durationMs = (event as any).durationMs as number;
        // Create and immediately end an OTel span carrying the full duration.
        const otelSpan = tracer.startSpan(span.name, {
          attributes: { ...span.attributes, 'dcp.duration_ms': durationMs },
          startTime: new Date(Date.now() - durationMs),
        });
        if (span.status === 'error') {
          otelSpan.setStatus({ code: api.SpanStatusCode.ERROR, message: span.error ?? 'error' });
        } else {
          otelSpan.setStatus({ code: api.SpanStatusCode.OK });
        }
        otelSpan.end();
        return;
      }

      if (event.type === 'metric') {
        const name = String((event as any).name);
        const value = Number((event as any).value);
        const labels = (event as any).labels as Record<string, string> | undefined;
        const attrs = labels ?? {};
        switch (name) {
          case 'sign_latency_ms':
            histogram('dcp.sign.latency_ms').record(value, attrs);
            counter('dcp.signatures.created').add(1, attrs);
            return;
          case 'verify_latency_ms':
            histogram('dcp.verify.latency_ms').record(value, attrs);
            counter('dcp.signatures.verified').add(1, attrs);
            return;
          case 'kem_latency_ms':
            histogram('dcp.kem.latency_ms').record(value, attrs);
            return;
          case 'checkpoint_latency_ms':
            histogram('dcp.checkpoint.latency_ms').record(value, attrs);
            return;
          case 'bundle_verify_ms':
            histogram('dcp.bundle_verify.latency_ms').record(value, attrs);
            counter('dcp.bundles.verified').add(1, attrs);
            return;
          case 'a2a_sessions_total':
            counter('dcp.a2a.sessions').add(1, attrs);
            return;
          default:
            // Unknown metric name — record on a generic histogram with the name as an attr.
            histogram('dcp.metric').record(value, { ...attrs, name });
            return;
        }
      }

      if (event.type === 'error') {
        const operation = String((event as any).operation);
        counter('dcp.errors').add(1, { operation });
        return;
      }
      // 'init' and anything else: no-op for OTel.
    } catch {
      // Telemetry must never break the application.
    }
  }

  async function shutdown(): Promise<void> {
    try {
      await metricReader.shutdown();
    } catch {
      /* ignore */
    }
    try {
      await tracerProvider.shutdown();
    } catch {
      /* ignore */
    }
  }

  return { handleEvent, shutdown };
}

File: sdks/rust/src/observability/otlp.rs
//! OTLP bridge — forwards `TelemetryEvent` values to OpenTelemetry traces &
//! metrics. Enabled via the `otlp` Cargo feature.

use std::collections::HashMap;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use opentelemetry::{
    global,
    metrics::{Counter, Histogram, Meter, MeterProvider as _},
    trace::{Tracer, TracerProvider as _},
    KeyValue,
};
use opentelemetry_otlp::WithExportConfig;
use opentelemetry_sdk::{
    metrics::{PeriodicReader, SdkMeterProvider},
    runtime,
    trace::{Sampler, TracerProvider},
    Resource,
};

use super::{TelemetryEvent, DCP_SDK_VERSION};

#[derive(Clone)]
pub struct OtlpHandles {
    pub tracer_provider: TracerProvider,
    pub meter_provider: SdkMeterProvider,
    pub tracer: opentelemetry_sdk::trace::Tracer,
    pub histograms: Arc<Histograms>,
    pub counters: Arc<Counters>,
}

pub struct Histograms {
    pub sign: Histogram<f64>,
    pub verify: Histogram<f64>,
    pub kem: Histogram<f64>,
    pub checkpoint: Histogram<f64>,
    pub bundle_verify: Histogram<f64>,
}

pub struct Counters {
    pub signatures_created: Counter<u64>,
    pub signatures_verified: Counter<u64>,
    pub bundles_verified: Counter<u64>,
    pub a2a_sessions: Counter<u64>,
    pub a2a_messages: Counter<u64>,
    pub errors: Counter<u64>,
}

pub fn init_bridge(service_name: &str, endpoint: Option<&str>) -> Result<OtlpHandles, String> {
    let endpoint = endpoint.unwrap_or("http://localhost:4318");

    let resource = Resource::new(vec![
        KeyValue::new("service.name", service_name.to_string()),
        KeyValue::new("service.version", DCP_SDK_VERSION.to_string()),
        KeyValue::new("sdk.language", "rust"),
    ]);

    let trace_endpoint = format!("{}/v1/traces", endpoint.trim_end_matches('/'));
    let metric_endpoint = format!("{}/v1/metrics", endpoint.trim_end_matches('/'));

    // Traces
    let tracer_provider = opentelemetry_otlp::new_pipeline()
        .tracing()
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .http()
                .with_endpoint(trace_endpoint),
        )
        .with_trace_config(
            opentelemetry_sdk::trace::Config::default()
                .with_sampler(Sampler::AlwaysOn)
                .with_resource(resource.clone()),
        )
        .install_batch(runtime::Tokio)
        .map_err(|e| format!("tracer install: {}", e))?;

    global::set_tracer_provider(tracer_provider.clone());
    let tracer = tracer_provider.tracer("dcp-ai");

    // Metrics
    let metric_exporter = opentelemetry_otlp::new_exporter()
        .http()
        .with_endpoint(metric_endpoint)
        .build_metrics_exporter(
            Box::new(opentelemetry_sdk::metrics::reader::DefaultAggregationSelector::new()),
            Box::new(opentelemetry_sdk::metrics::reader::DefaultTemporalitySelector::new()),
        )
        .map_err(|e| format!("metric exporter: {}", e))?;

    let reader = PeriodicReader::builder(metric_exporter, runtime::Tokio).build();
    let meter_provider = SdkMeterProvider::builder()
        .with_reader(reader)
        .with_resource(resource)
        .build();
    global::set_meter_provider(meter_provider.clone());
    let meter: Meter = meter_provider.meter("dcp-ai");

    let histograms = Arc::new(Histograms {
        sign: meter.f64_histogram("dcp.sign.latency_ms").init(),
        verify: meter.f64_histogram("dcp.verify.latency_ms").init(),
        kem: meter.f64_histogram("dcp.kem.latency_ms").init(),
        checkpoint: meter.f64_histogram("dcp.checkpoint.latency_ms").init(),
        bundle_verify: meter.f64_histogram("dcp.bundle_verify.latency_ms").init(),
    });
    let counters = Arc::new(Counters {
        signatures_created: meter.u64_counter("dcp.signatures.created").init(),
        signatures_verified: meter.u64_counter("dcp.signatures.verified").init(),
        bundles_verified: meter.u64_counter("dcp.bundles.verified").init(),
        a2a_sessions: meter.u64_counter("dcp.a2a.sessions").init(),
        a2a_messages: meter.u64_counter("dcp.a2a.messages").init(),
        errors: meter.u64_counter("dcp.errors").init(),
    });

    Ok(OtlpHandles {
        tracer_provider,
        meter_provider,
        tracer,
        histograms,
        counters,
    })
}

fn to_kv(labels: &HashMap<String, String>) -> Vec<KeyValue> {
    labels
        .iter()
        .map(|(k, v)| KeyValue::new(k.clone(), v.clone()))
        .collect()
}

pub fn forward(handles: &OtlpHandles, event: &TelemetryEvent) {
    match event {
        TelemetryEvent::Metric { name, value, labels, .. } => {
            let kv = to_kv(labels);
            match name.as_str() {
                "sign_latency_ms" => handles.histograms.sign.record(*value, &kv),
                "verify_latency_ms" => handles.histograms.verify.record(*value, &kv),
                "kem_latency_ms" => handles.histograms.kem.record(*value, &kv),
                "checkpoint_latency_ms" => handles.histograms.checkpoint.record(*value, &kv),
                "bundle_verify_ms" => handles.histograms.bundle_verify.record(*value, &kv),
                _ => {}
            }
        }
        TelemetryEvent::Counter { name, labels, .. } => {
            let kv = to_kv(labels);
            match name.as_str() {
                "signatures_created" => handles.counters.signatures_created.add(1, &kv),
                "signatures_verified" => handles.counters.signatures_verified.add(1, &kv),
                "bundles_verified" => handles.counters.bundles_verified.add(1, &kv),
                "a2a_sessions" => handles.counters.a2a_sessions.add(1, &kv),
                "a2a_messages" => handles.counters.a2a_messages.add(1, &kv),
                _ => {}
            }
        }
        TelemetryEvent::Error { operation, .. } => {
            handles
                .counters
                .errors
                .add(1, &[KeyValue::new("operation", operation.clone())]);
        }
        TelemetryEvent::Span { span, .. } => {
            use opentelemetry::trace::{Span, SpanKind, Status};
            let mut s = handles
                .tracer
                .span_builder(span.name.clone())
                .with_kind(SpanKind::Internal)
                .with_start_time(system_time_from_ms(span.start_time_ms))
                .start(&handles.tracer);
            for (k, v) in span.attributes.0.iter() {
                use super::AttrValue::*;
                let key = k.clone();
                let kv = match v {
                    Str(s) => KeyValue::new(key, s.clone()),
                    Int(i) => KeyValue::new(key, *i),
                    Float(f) => KeyValue::new(key, *f),
                    Bool(b) => KeyValue::new(key, *b),
                };
                s.set_attribute(kv);
            }
            if span.status == super::SpanStatus::Error {
                s.set_status(Status::error(span.error.clone().unwrap_or_default()));
            } else {
                s.set_status(Status::Ok);
            }
            if let Some(end_ms) = span.end_time_ms {
                s.end_with_timestamp(system_time_from_ms(end_ms));
            } else {
                s.end();
            }
        }
        TelemetryEvent::Init { .. } => {}
    }
}

fn system_time_from_ms(ms: u64) -> SystemTime {
    UNIX_EPOCH + std::time::Duration::from_millis(ms)
}

pub fn shutdown(handles: OtlpHandles) {
    let _ = handles.tracer_provider.shutdown();
    let _ = handles.meter_provider.shutdown();
}
